[Inlined Callstack Fix] Fix inlined callstack for blocks of the node.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57481 [Pytorch Delegated Backend] Save function name in debug info
* #57441 [Pytorch Delegated Backend] Save operator name and function name in
* #57156 [Pytorch Delegated Backend] Add python binding for generate_debug_handles
* #55462 [Pytorch Backend delegation] Add api for backend lowering to query debug
* **#56562 [Inlined Callstack Fix] Fix inlined callstack for blocks of the node.**
* #55252 [Pytorch Mobile DebugInfo Serialization] Save debug handles for all instructions.
* #55062 [Pytorch, Mobile] Serialize inlined callstack pointer with debug handle.
* #54284 [PyTorch, Mobile] Serialization format change for source range

Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

kimishpatel committed:
[Inlined Callstack Fix] Fix inlined callstack for blocks of the node.

Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "[Inlined Callstack Fix] Fix inlined callstack for blocks of the node."


Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Inlined Callstack Fix] Fix inlined callstack for blocks of the node."


Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Inlined Callstack Fix] Fix inlined callstack for blocks of the node."


Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Inlined Callstack Fix] Fix inlined callstack for blocks of the node."


Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Inlined Callstack Fix] Fix inlined callstack for blocks of the node."


Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

[ghstack-poisoned]

ZolotukhinM commented on test/cpp/jit/test_misc.cpp:
```diff
@@ -2039,6 +2040,63 @@ def c(x):
+  auto graph = c.get_method("forward").graph();
+  torch::jit::Inline(*graph);
+  std::stringstream add_ss, mul_ss;
+  for (Node* n : graph->nodes()) {
+    if (n->kind() == prim::If) {
+      for (Block* block : n->blocks()) {
+        for (Node* if_node : block->nodes()) {
+          if (if_node->kind() == aten::add) {
```
Nit: this could be factored out as a function `findNodeOfGivenKind` or something like this.

ZolotukhinM replied:
Btw, you could try to use subgraph matcher for this, but i'm not sure that with the required boilerplate code it would be beneficial.

ZolotukhinM replied:
This comment still applies.

kimishpatel replied:
True. But looking at this, it looks like it is just a 10ish lines of code without much resuse. If more similar code is added I suppose we can refactor. So for now I will leave it as is unless you have a strong opinion about it.

ZolotukhinM commented on torch/csrc/jit/ir/ir.cpp:
```diff
@@ -1929,6 +1929,36 @@ at::ArrayRef<Value*> createTupleUnpack(Value* v) {
+  for (auto n : b->nodes()) {
+    auto new_node_cs = n->callstack();
+
+    InlinedCallStack* raw_callstack_ptr =
+        new_node_cs ? new_node_cs->get() : nullptr;
+
+    if (!new_cs_entires.count(raw_callstack_ptr)) {
+      if (new_node_cs) {
+        new_cs_entires[raw_callstack_ptr] =
+            c10::make_intrusive<InlinedCallStack>(
+                *new_node_cs, callee, to_replace->sourceRange(), m_info);
+      } else {
+        new_cs_entires[raw_callstack_ptr] =
+            c10::make_intrusive<InlinedCallStack>(
+                callee, to_replace->sourceRange(), m_info);
+      }
+    }
```
This seem to duplicate the code from `inlineCallTo` almost identically. Would it be possible to get rid of that duplication by calling `inlineCallStackOfBlock` there?

kimishpatel replied:
Yes. I was almost gonna do that, but wanted to move too fast. Let me refactor it.

kimishpatel replied:
> Also, why is it possible that !new_cs_entires.count(raw_callstack_ptr) is true but if (new_node_cs) is false?

So say you are inlining:
```
class B(..):
  def forward(self, x):
     return x + 3
class A(..):
  def forward(self, y):
    return self.b.forward(y)
```
Then when you are inlining A's forward method you replace
`%something = prim::CallMethod(...)` with `%something = aten::add(..`
This aten::add node is created by cloning B's forward method. At that time aten::add node's cs is null. And new_cs_entires may not have an entry for raw_callstack_ptr = nullptr so `if (!new_cs_entires.count(raw_callstack_ptr)) ` would be true.

ZolotukhinM reviewed (commented):
Overall this looks reasonable, I have a couple of question inline:

raziel commented on test/cpp/jit/test_misc.cpp:
```diff
@@ -2039,6 +2040,63 @@ def c(x):
+  ASSERT_NE(mul_ss.str().find("line 6"), std::string::npos);
+  ASSERT_NE(
+      mul_ss.str().find("return self.A0.forward(x, y, z)"), std::string::npos);
+  ASSERT_NE(add_ss.str().find("return x * y"), std::string::npos);
```
why is return x * y in add_ss and not mul_ss?

kimishpatel replied:
Oh thats a typo. Althoug the reason this still passes is because I think source range prints a few lines (2 I think) around the actual source.

kimishpatel replied:
I really thought I did.

kimishpatel committed:
Update on "[Inlined Callstack Fix] Fix inlined callstack for blocks of the node."


Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Inlined Callstack Fix] Fix inlined callstack for blocks of the node."


Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

[ghstack-poisoned]

raziel commented on torch/csrc/jit/ir/ir.cpp:
```diff
@@ -1930,6 +1930,53 @@ at::ArrayRef<Value*> createTupleUnpack(Value* v) {
+
+void inlineCallStackOfNode(
+    Node* new_node,
+    std::unordered_map<InlinedCallStack*, InlinedCallStackPtr>& new_cs_entires,
```
So the name is really new_cs_entires? not new_cs_entries

kimishpatel replied:
LOL. OK thanks for that catch. I missed that somehow.

kimishpatel committed:
Update on "[Inlined Callstack Fix] Fix inlined callstack for blocks of the node."


Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Inlined Callstack Fix] Fix inlined callstack for blocks of the node."


Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Inlined Callstack Fix] Fix inlined callstack for blocks of the node."


Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Inlined Callstack Fix] Fix inlined callstack for blocks of the node."


Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

[ghstack-poisoned]

ZolotukhinM commented on test/cpp/jit/test_misc.cpp:
```diff
@@ -2126,6 +2127,64 @@ def c(x):
+      return self.A0.forward(x, y, z) + self.B0.forward(x)
+  )");
+
+  auto graph = c.get_method("forward").graph();
+  torch::jit::Inline(*graph);
```
If you use `optimized_graph()` method you won't need to call `Inline` directly (similarly to how it was done in existing tests).

ZolotukhinM commented on test/cpp/jit/test_misc.cpp:
```diff
@@ -31,6 +31,7 @@
 #include <torch/csrc/jit/passes/graph_fuser.h>
 #include <torch/csrc/jit/passes/guard_elimination.h>
 #include <torch/csrc/jit/passes/inline_autodiff_subgraphs.h>
+#include <torch/csrc/jit/passes/inliner.h>
```
This would be unnecessary if we use `optimized_graph` instead of just `graph`.

ZolotukhinM reviewed (approved):
LGTM, only a couple of minor remarks from me.

kimishpatel committed:
Update on "[Inlined Callstack Fix] Fix inlined callstack for blocks of the node."


Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Inlined Callstack Fix] Fix inlined callstack for blocks of the node."


Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Inlined Callstack Fix] Fix inlined callstack for blocks of the node."


Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Inlined Callstack Fix] Fix inlined callstack for blocks of the node."


Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Inlined Callstack Fix] Fix inlined callstack for blocks of the node."


Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Inlined Callstack Fix] Fix inlined callstack for blocks of the node."


Summary:
Earlier inlined callstack was annotated only nodes. This left out nodes
such as If which have block of nodes. These nodes should also be updated
similarly.

Test Plan:
Added test in test_misc

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27902516](https://our.internmc.facebook.com/intern/diff/D27902516)

[ghstack-poisoned]
